Switch to different finalisation mechanism

jacob-hughes opened the pull request:
This is the first of a couple of a couple PRs to make finalisation in Alloy faster. See the commit messages for details.

ltratt commented on library/std/src/gc.rs:
```diff
@@ -399,6 +392,37 @@ impl<T> Gc<T> {
+
+#[cfg(profile_gc)]
+#[derive(Debug)]
+pub struct FinalizerInfo {
```
I guess this is more `FinalizersInfo` or `GCProfileInfo` or something like that?

jacob-hughes replied:
Yes good point. But this existed before, this commit has just moved it around. Shall I rename it?

ltratt replied:
I think renaming it would be a good idea, but you could do it in a subsequent PR. Up to you.

ltratt commented:
In 4d3bb73's commit message:

> The disclaim approach will only work with off-thread finalisation because it
runs finalisers during the STW pause where the allocator lock is already
acquired, thus finalisers run on the mutator would not be able to perform
anything which interacts with the heap. It is therefore not possible to compare
the performance of this approach with the performance of single-threaded
allocation.

I don't understand the last sentence, as I think a) the paragraph following it does seem to give a performance comparison vs. the previous version b) we have compared that previous version with single-threaded?

ltratt commented on library/std/src/gc.rs:
```diff
@@ -72,6 +74,29 @@ pub const MIN_ALIGN: usize = 8;
+pub struct GcFinalizedAllocator;
+
+use bdwgc::FinalizerClosure;
+static FINALIZER_CLOSURE: FinalizerClosure =
```
Is this a dummy value of sorts? Can we document that.

jacob-hughes replied:
Think of this as the entry point to the finaliser. Boehm's `finalized_malloc` function takes a struct with two fields: a pointer to the fn pointer, and some additional metadata (which we don't use). Since this is re-used for every finalisable object we make this a single static value to save us from having to make a new struct for each object. I'll document this.

ltratt replied:
Right, let's document that then.

jacob-hughes replied:
How about something like this? 8fbd2902e45817e6e156e6843a805f4e024db2c5

ltratt commented on library/std/src/gc.rs:
```diff
@@ -175,11 +201,76 @@ pub fn thread_registered() -> bool {
+//
+// This is considered a root for the Gc because the 'to-be-finalised' objects are otherwise
+// unreachable and must be kept alive until their finaliser has run.
+static FIN_Q: Mutex<VecDeque<Gc<()>>> = Mutex::new(VecDeque::new());
```
Doesn't this mean we're doing heap allocations *during* GC? I'm also not sure about the performance properties of `VecDeque`, particularly as I guess that we often shove huge numbers of objects in here. Update: ah, I see on line 225 that we ensure we don't enlarge it! What happens then?!

I also guess that we're going to be heavily punished for the `Mutex` here, as presumably parallel GC ends up locking/unlocking this heavily during particular parts of the GC cycle. I don't know if there's a good way of measuring that.

I don't think we should do this in this PR but I think we could fairly easily do this in the next PR:

1. Each GC thread has a `thread_local` queue of objects to finalise. That could be a `VecDeque` but (because of the "what happens if we run out of space in the `VecDeque` problem) I think that might better be an `mmap`ed block (or similar) under our control.
2. At the end of the GC cycle (or when that local queue is full) we flush it to a global "queue of queues", which is something like `Mutex<Vec<FinalisationQueue>>`.

The aim here being that our number of locks goes from O(number_of_objects_to_finalise) to O(number_of_GC_threads).

jacob-hughes replied:
> Update: ah, I see on line 225 that we ensure we don't enlarge it! What happens then?!

We check whether we need to enlarge the queue on each allocation because this is the only time it would be safe to do so. I'll be the first to admit I don't like this, but I couldn't see another a way around accidentally growing the ring buffer while inside the STW pause.

> I also guess that we're going to be heavily punished for the Mutex here, as presumably parallel GC ends up locking/unlocking this heavily during particular parts of the GC cycle. I don't know if there's a good way of measuring that.

Profile data still shows that we spend a lot of time in `futex_wait` syscalls -- especially from within the GC. However, we're punished for this `Mutex` more because of contention between the finaliser thread and user mutator threads. The more _concurrent_ mutator (application) threads we have, the worse this gets. However,  this does not affect parallel collector threads (which do marking) because they don't lock/unlock this mutex (and Rust mutexes don't heap allocate anymore).

I agree your suggested approach is better. It's similar to what I had lined up (but still suffers from the `VecDeque` grow problem). I'm not sure what you mean by this though:

>  I think that might better be an mmaped block (or similar) under our control.

Do you mean `mmap` a block of pointers to objects that need finalising with the system allocator (i.e. outside of Boehm)? We'd still need to grow this block dynamically, and we also need Boehm to know about it so that it can treat it as a source of roots.

> At the end of the GC cycle (or when that local queue is full) we flush it to a global "queue of queues", which is something like Mutex<Vec<FinalisationQueue>>.

Right, so we batch up finalisation in pre-defined chunks that get siphoned off to the finaliser thread once they reach some predetermined size?

ltratt replied:
>We check whether we need to enlarge the queue on each allocation because this is the only time it would be safe to do so.

Right, so this isn't something we ideally want to do in the fast path.

> Right, so we batch up finalisation in pre-defined chunks that get siphoned off to the finaliser thread once they reach some predetermined size?

Yep, that's the key idea!

>Do you mean mmap a block of pointers to objects that need finalising using system allocator (i.e. outside of Boehm)?

Assuming we can teach Boehm about the `mmap`ed block of pointers somehow, that would be my suggestion. But there are other conceivable ways which achieve the same outcome.

> We'd still need to grow this block dynamically, and we also need Boehm to know about it so that it can treat it as a source of roots.

Yes.

jacob-hughes replied:
Ok, with this + https://github.com/softdevteam/alloy/pull/114#discussion_r1544803488 I think we might be able to eliminate synchronisation around the `FIN_Q`. Ok to do this in the next PR?

ltratt replied:
Works for me.

ltratt commented on library/std/src/gc.rs:
```diff
@@ -175,11 +201,76 @@ pub fn thread_registered() -> bool {
+
+pub fn process_fin_q() {
+    loop {
+        let object = FIN_Q.lock().unwrap().pop_front();
```
This is interesting: I *think* we only have a 1 finalisation thread, right? But that thread still has to acquire a lock to do finalisation even though *we* know that no other thread can have gained that lock (outside of GC time).

If I've got that right, it suggests that we also want to *move* the "queue of queues" to the finalisation thread, so that it doesn't have to do any locking.

jacob-hughes replied:
> This is interesting: I think we only have a 1 finalisation thread, right? 

Correct

> But that thread still has to acquire a lock to do finalisation even though we know that no other thread can have gained that lock (outside of GC time).

But the `FIN_Q` is like a producer / consumer, so we need to lock because this could be popping from the queue at the same time that new allocations on the mutator push to the queue. Maybe I've misunderstood your suggestion.

ltratt replied:
>But the FIN_Q is like a producer / consumer, so we need to lock because this could be popping from the queue at the same time that new allocations on the mutator push to the queue. Maybe I've misunderstood your suggestion.

The finaliser thread will be stopped during GC so at run-time I don't think it can ever actually race? But we will stay pay the lock/unlock costs even though we're not really doing anything that could ever race...

jacob-hughes replied:
> The finaliser thread will be stopped during GC so at run-time I don't think it can ever actually race? But we will stay pay the lock/unlock costs even though we're not really doing anything that could ever race...

Doh... You're right! This hadn't occurred to me. If I can then malloc / mmap this queue by side-stepping the Boehm allocator, I might be able to grow it inside the STW pause and remove locking altogether.

ltratt commented:
@jacob-hughes Please squash, but in so doing can you also address the question in https://github.com/softdevteam/alloy/pull/114#issuecomment-2026833017 please?

jacob-hughes commented:
Squashed. 

> I don't understand the last sentence, as I think a) the paragraph following it does seem to give a performance comparison vs. the previous version b) we have compared that previous version with single-threaded?

I've removed this sentence because it was a bit confusing. You're right, all I was trying to say was that a direct comparison with single-threaded finalisation is not possible. We can infer that though from the comparison between the first off-thread finalisation approach in #a58c2b8

ltratt commented:
> You're right, all I was trying to say was that a direct comparison with single-threaded finalisation is not possible. We can infer that though from the comparison between the first off-thread finalisation approach in #a58c2b8

Maybe I'm an idiot, but a direct comparison *is* possible, but it's just awkward to do?

ltratt commented:
OK, that makes more sense --- thanks!

jacob-hughes commented:
Fixed here: 324150c7b1dffe9a284f5e43cfb4a34bfaf92e39

ltratt commented:
Please squash.

jacob-hughes commented:
I think this should stay as a separate commit that gets squashed when we rebase with upstream rustc. It's not related to this work and it's useful to keep track of.

jacob-hughes commented:
Fixed here: 78bc4a0d9ae21e7b739c02c62ddda422e130582e (Again I'm not sure this one should be squashed either)

jacob-hughes commented:
This failure is a little bit more tricky, I'm debugging it in tryci. However, I can use it as an opportunity to change to `debian:latest` in the buildbot config and then re-squash it later if you're ok with that?

ltratt commented:
"bookworm" and "latest" are currently identical, so you can push a new commit to that effect, but please don't squash it yet.

jacob-hughes commented:
Can I squash this to remove the top two commits?

ltratt commented:
Please squash.

jacob-hughes commented:
Squashed (please hold off on merging for now)

jacob-hughes commented:
Closing in favour of https://github.com/softdevteam/alloy/pull/117
